contracts/README.md

```diff
@@ -3,15 +3,396 @@
 
 # ENSv2 Contracts
 
-Currently this repository hosts Proof-of-Concept contracts for ENSv2. See the [ENSv2 design doc](http://go.ens.xyz/ensv2) for details of the system architecture.
+This repository hosts the smart contracts for ENSv2 (Ethereum Name Service version 2), a next-generation naming system designed for scalability and cross-chain functionality. For comprehensive architectural details, see the [ENSv2 design doc](http://go.ens.xyz/ensv2).
 
-At present the following contracts are implemented:
+## Table of Contents
 
-- [RegistryDatastore](src/common/RegistryDatastore.sol) &mdash; an implementation of the registry datastore defined in the design doc. All registry contracts must use a singleton instance of the datastore for storage of subregistry and resolver addresses.
-- [ERC1155Singleton](src/common/ERC1155Singleton.sol) &mdash; an implementation of the ERC1155 standard that permits only a single token per token ID. This saves on gas costs for storage while also permitting easy implementation of an `ownerOf` function.
-- [PermissionedRegistry](src/common/PermissionedRegistry.sol) &mdash; an implementation of the v2 registry.
-- [UniversalResolver](src/universalResolver/UniversalResolver.sol) &mdash; onchain ENSv2 resolution.
-- [ETHFallbackResolver](src/L1/ETHFallbackResolver.sol) &mdash; crosschain resolver that combines mainnet v2 (ejected), mainnet v1 (unmigrated), and Namechain v2.
+- [Overview](#overview)
+- [Architecture](#architecture)
+  - [Core Concepts](#core-concepts)
+  - [Token ID System](#token-id-system)
+  - [Access Control](#access-control)
+- [Contract Documentation](#contract-documentation)
+  - [Registry System](#registry-system)
+  - [L2 Components](#l2-components)
+  - [L1 Components](#l1-components)
+  - [Cross-Chain Bridge](#cross-chain-bridge)
+  - [Resolution](#resolution)
+- [Getting started](#getting-started)
+- [Running the Devnet](#running-the-devnet)
+- [Miscellaneous](#miscellaneous)
+
+## Overview
+
+ENSv2 transitions from a flat registry to a hierarchical system that enables:
+
+- **L2 Scalability**: .eth names registered and managed on Layer 2 with reduced gas costs
+- **Flexible Ownership**: Custom registry implementations for different ownership models
+- **Cross-Chain Support**: Seamless name resolution across L1 and L2
+- **Backward Compatibility**: Unmigrated ENSv1 names continue to function
+- **Enhanced Security**: Names can be "ejected" to L1 for maximum security guarantees
+
+### Key Features
+
+1. **Hierarchical Registries**: Each name has its own registry contract managing its subdomains
+2. **Canonical ID System**: Stable storage identifiers enabling token regeneration
```

Comment:
Doesn't quite explain what it's for. Canonical internal token ID enables for external token ID to be changed but still map to the same internal data.